linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -4683,4 +4683,191 @@ private void flush() {
       }
     };
   }
+
+  /**
+   * Merge Union Enumerable.
+   * Performs a union (or union all) of all its inputs (which must be already sorted),
+   * respecting the order.
+   *
+   * @param sources input enumerables (must be already sorted)
+   * @param sortKeySelector sort key selector
+   * @param sortComparator sort comparator to decide the next item
+   * @param all whether duplicates will be considered or not
+   * @param equalityComparer {@link EqualityComparer} to control duplicates,
+   *                         only used if {@code all} is {@code false}
+   * @param <TSource> record type
+   * @param <TKey> sort key
+   */
+  public static <TSource, @Nullable TKey> Enumerable<TSource> mergeUnion(
+      List<Enumerable<TSource>> sources,
+      Function1<TSource, @Nullable TKey> sortKeySelector,
+      Comparator<TKey> sortComparator,
+      boolean all,
+      EqualityComparer<TSource> equalityComparer) {
+    return new AbstractEnumerable<TSource>() {
+      @Override public Enumerator<TSource> enumerator() {
+        return new MergeUnionEnumerator<>(
+            sources,
+            sortKeySelector,
+            sortComparator,
+            all,
+            equalityComparer);
+      }
+    };
+  }
+
+  /**
+   * Performs a union (or union all) of all its inputs (which must be already sorted),
+   * respecting the order.
+   * @param <TSource> record type
+   * @param <TKey> sort key
+   */
+  @SuppressWarnings("unchecked")
+  private static class MergeUnionEnumerator<TSource, @Nullable TKey>
+      implements Enumerator<TSource> {
+    private final List<Enumerator<TSource>> inputs;
+    private final TSource[] currentInputsValues;
+    private final Function1<TSource, @Nullable TKey> sortKeySelector;
+    private final Comparator<@Nullable TKey> sortComparator;
+    private TSource currentValue;
+    private int activeInputs;
+
+    // Set to control duplicates, only used if "all" is false
+    private final @Nullable Set<Wrapped<TSource>> processed;
```

Comment:
The inputs are sorted, so you do not need `Set` here. A single `TSource previous` record is enough, then you could compare `previous` with a new entry to tell if the next one should be emitted or not.

Reply:
You're right, I will work in the `TSource previous` approach

Reply:
> Input1 = Emp(10, Bob), Emp(20, Joe), Emp(20, Greg)

What is the collation?

Reply:
collation is the first field (deptId), so $0.

Reply:
Oh, so you mean that `sort-merge union` keeps the common collation while merging the results, however, it should eliminate duplicates no matter the order.

Then the `set` is indeed required. However, the set can be skipped in case all the columns participate in the collation (e.g. `order by id, name` ).

Reply:
Exactly.

Reply:
On the other hand, the set can be invalidated as the collation key changes which would reduce memory consumption. WDYT?

Reply:
It's a good idea, I'll work on that.

Reply:
Optimization committed